pkg/gadgets/trace/byob/tracer/tracer.go

```diff
@@ -0,0 +1,346 @@
+				errVerifier)
+		}
+		return fmt.Errorf("failed to create BPF collection: %w", err)
+	}
+
+	m := t.collection.Maps[t.printMap]
+	if m.Type() == ebpf.RingBuf {
+		t.ringbufReader, err = ringbuf.NewReader(t.collection.Maps[t.printMap])
+	} else if m.Type() == ebpf.PerfEventArray {
+		t.perfReader, err = perf.NewReader(t.collection.Maps[t.printMap], gadgets.PerfBufferPages*os.Getpagesize())
+	} else {
+		return fmt.Errorf("unsupported BPF map type: %s", m.Type())
+	}
+	if err != nil {
+		return fmt.Errorf("failed to create BPF map reader: %w", err)
+	}
+	go t.run()
+
+	// Attach programs
+	for progName, p := range t.spec.Programs {
+		if p.Type == ebpf.Kprobe && strings.HasPrefix(p.SectionName, "kprobe/") {
+			l, err := link.Kprobe(p.AttachTo, t.collection.Programs[progName], nil)
+			if err != nil {
+				return fmt.Errorf("failed to attach BPF program %q: %w", progName, err)
+			}
+			t.links = append(t.links, l)
+		} else if p.Type == ebpf.Kprobe && strings.HasPrefix(p.SectionName, "kretprobe/") {
+			l, err := link.Kretprobe(p.AttachTo, t.collection.Programs[progName], nil)
+			if err != nil {
+				return fmt.Errorf("failed to attach BPF program %q: %w", progName, err)
+			}
+			t.links = append(t.links, l)
+		}
+	}
+
+	return nil
+}
+
+func (t *Tracer) run() {
+	d := t.decoderFactory()
+	ctx := context.TODO()
+	for {
+		var rawSample []byte
+
+		if t.ringbufReader != nil {
+			record, err := t.ringbufReader.Read()
+			if err != nil {
+				if errors.Is(err, ringbuf.ErrClosed) {
+					// nothing to do, we're done
+					return
+				}
+
+				msg := fmt.Sprintf("Error reading ring buffer: %s", err)
+				t.eventCallback(types.Base(eventtypes.Err(msg)))
+				return
+			}
+			rawSample = record.RawSample
+		} else if t.perfReader != nil {
+			record, err := t.perfReader.Read()
+			if err != nil {
+				if errors.Is(err, perf.ErrClosed) {
+					return
+				}
+
+				msg := fmt.Sprintf("Error reading perf ring buffer: %s", err)
+				t.eventCallback(types.Base(eventtypes.Err(msg)))
+				return
+			}
+
+			if record.LostSamples != 0 {
+				msg := fmt.Sprintf("lost %d samples", record.LostSamples)
+				t.eventCallback(types.Base(eventtypes.Warn(msg)))
+				continue
+			}
+			rawSample = record.RawSample
+		} else {
+			msg := "Error using reader for ring buffer"
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
+
+		if uint32(len(rawSample)) < t.mapSizes[t.printMap] {
+			msg := fmt.Sprintf("Error reading ring buffer: len(RawSample)=%d!=%d",
+				len(rawSample),
+				t.mapSizes[t.printMap])
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
+
+		// FIXME: DecodeBtfBinary has a bug with non-NULL-terminated strings.
+		// For now, ensure the problem does not happen in ebpf
+
+		result, err := d.DecodeBtfBinary(ctx, t.valueStruct, rawSample[:t.mapSizes[t.printMap]])
+		if err != nil {
+			msg := fmt.Sprintf("Error decoding btf: %s", err)
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
```

Comment:
`DecodeBtfBinary` is really amazing! I think we can integrate this with the existing columns library to make this more powerful. A couple of ideas I have:

1. Extend the columns library to allow creating it from BTF information: Similarly to [NewColumns](https://github.com/inspektor-gadget/inspektor-gadget/blob/3030c5267623071996d6b3e2d434ba939480626b/pkg/columns/columns.go#L51), we could have a method that creates the columns from a `btf.Type,`. We won't support all features that we support for golang annotated types, but we can use the information in the BTF to pretty-print the data. Later on, we can explore if there is a way to include information about the column in the eBPF code as we do in the golang code. 
2. The parser has to be gathered in another moment. Given that the columns are created a runtime by the gadget, the parser used to print the data should be returned after the gadget starts running. @flyth any thoughts on this one?